### PR TITLE
feat: add shell command to connect to an application

### PR DESCRIPTION
**Pull request by benamib**

### CODE REVIEW

File: pkg/shell.go [hunk 1]
```diff
@@ -11,16 +20,102 @@ type ShellRequest struct {
...
-	fmt.Println(wsURL)
+	//wsURL := fmt.Sprintf("wss://ws.qovery.com/shell/exec?application=%s&environment=%s&project=%s&organization=%s&cluster=%s",
+	//	req.ApplicationID,
+	//	req.EnvironmentID,
+	//	req.ProjectID,
+	//	req.OrganizationID,
+	//	req.ClusterID,
+	//)
+
+	u, err := url.Parse("wss://ws.qovery.com/shell/exec?organization=141c07c8-0dd9-4623-983b-3fdd61867255&cluster=a8ad0659-b8f2-4c83-ad77-092e97bb2cae&project=5cb54778-b8f2-4c83-ad77-092e97bb2cae&environment=2df1d928-0dd9-4623-983b-3fdd61867255&application=e44c0adc-b8f2-4c83-ad77-092e97bb2cae")
```
Comment (mzottola) [inline pkg/shell.go:35]: to remove

File: pkg/shell.go [hunk 2]
```diff
@@ -11,16 +20,102 @@ type ShellRequest struct {
...
+	if err != nil {
+		log.Fatal(err)
+	}
+
+	log.Printf("connecting to %s", u.String())
```
Comment (mzottola) [inline pkg/shell.go:40]: to change

File: pkg/shell.go [hunk 3]
```diff
@@ -11,16 +20,102 @@ type ShellRequest struct {
...
+
+	token, err := utils.GetAccessToken()
+	if err != nil {
+		log.Fatal(err)
+	}
+	headers := http.Header{
+		"Authorization": {fmt.Sprintf("Bearer %s", token)},
+	}
+	c, _, err := websocket.DefaultDialer.Dial(u.String(), headers)
```
Comment (mzottola) [inline pkg/shell.go:49]: to rename `c` in `websocketConnection` or other

File: pkg/shell.go [hunk 4]
```diff
@@ -11,16 +20,102 @@ type ShellRequest struct {
...
+		select {
+		case <-done:
+			return
+		case b := <-stdIn:
+			err := c.WriteMessage(websocket.TextMessage, b)
+			if err != nil {
+				log.Println("write:", err)
+				return
+			}
+		case <-interrupt:
```
Comment (mzottola) [inline pkg/shell.go:104]: interrupt is done using CTRL+D, to see if we can fix to CTRL+C

File: pkg/shell.go [hunk 5]
```diff
@@ -11,16 +20,102 @@ type ShellRequest struct {
...
+			log.Println("interrupt")
+
+			// Cleanly close the connection by sending a close message and then
+			// waiting (with timeout) for the server to close the connection.
+			err := c.WriteMessage(websocket.CloseMessage, websocket.FormatCloseMessage(websocket.CloseNormalClosure, ""))
```
Comment (mzottola) [inline pkg/shell.go:109]: To see with @erebe if we need to change this behavior

File: pkg/shell.go [hunk 6]
```diff
@@ -11,16 +20,102 @@ type ShellRequest struct {
...
+		for {
+			_, err := currentConsole.Read(b)
+			if err != nil {
+				log.Fatal("read", err)
+			}
+			stdIn <- b
+		}
+	}()
+
+	ticker := time.NewTicker(time.Second)
```
Comment (mzottola) [inline pkg/shell.go:91]: to remove

File: pkg/shell.go [hunk 7]
```diff
@@ -0,0 +1,115 @@
...
+		if _, err = currentConsole.Write(msg); err != nil {
+			log.Error("error while writing in console:", err)
+			return
+		}
+	}
+}
+
+func readUserConsole(currentConsole console.Console, stdIn chan []byte, done chan struct{}) {
+	defer close(done)
+	buffer := make([]byte, 4096)
```
Comment (erebe) [inline pkg/shell.go:106]: OPT: If possible to have it in parameter/config of the function

**Review by erebe — APPROVED** [review]

**Comment by erebe** [comment]
Do we send in the end the "stty -echo" ?

File: pkg/shell.go (hunk 7 above)
Comment (benamib) [reply pkg/shell.go:106]: I can put this as a const yes 👍

**Review by mzottola — APPROVED** [review]

File: pkg/shell.go [hunk 8]
```diff
@@ -0,0 +1,117 @@
...
+	return wsConn, nil
+}
+
+func readWebsocketConnection(wsConn *websocket.Conn, currentConsole console.Console, done chan struct{}) {
+	defer close(done)
+	for {
+		_, msg, err := wsConn.ReadMessage()
+		if err != nil {
+			if e, ok := err.(*websocket.CloseError); ok {
+				errSplit := strings.Split(e.Error(), ":")
```
Comment (erebe) [inline pkg/shell.go:92]: Can you remove this split.
It was an error on my side, I was returning more that the error message in the close frame.

So you should not have to do that on your side

**Pull request merged**